src/next_array.rs

```diff
@@ -0,0 +1,80 @@
+use core::mem::MaybeUninit;
+
+/// Helper struct to build up an array element by element.
+struct ArrayBuilder<T, const N: usize> {
+    arr: [MaybeUninit<T>; N],
+    i: usize
+}
+
+impl<T, const N: usize> ArrayBuilder<T, N> {
+    pub fn new() -> Self {
+        Self { arr: maybe_uninit::uninit_array(), i: 0 }
+    }
+    
+    pub unsafe fn push_unchecked(&mut self, x: T) {
+        debug_assert!(self.i < N);
+        *self.arr.get_unchecked_mut(self.i) = MaybeUninit::new(x);
```

Comment:
Needs a safety comment in the form:
```suggestion
        // SAFETY: By contract on the caller, the safety condition on `get_unchecked_mut` that BLAH BLAH BLAH is satisfied.
        *self.arr.get_unchecked_mut(self.i) = MaybeUninit::new(x);
```